consensus/polybft/validators_snapshot.go

```diff
@@ -214,29 +221,96 @@ func (v *validatorsSnapshotCache) storeSnapshot(block uint64, snapshot AccountSe
 // Cleanup cleans the validators cache in memory and db
 func (v *validatorsSnapshotCache) cleanup() error {
 	if len(v.snapshots) >= validatorSnapshotLimit {
-		latestBlock := uint64(0)
+		latestEpoch := uint64(0)
 
 		for e := range v.snapshots {
-			if e > latestBlock {
-				latestBlock = e
+			if e > latestEpoch {
+				latestEpoch = e
 			}
 		}
 
-		block := latestBlock
-		cache := make(map[uint64]AccountSet, numberOfSnapshotsToLeaveInMemory)
+		startEpoch := latestEpoch
+		cache := make(map[uint64]*validatorSnapshot, numberOfSnapshotsToLeaveInMemory)
 
 		for i := 0; i < numberOfSnapshotsToLeaveInMemory; i++ {
-			if snapshot, exists := v.snapshots[block]; exists {
-				cache[block] = snapshot
+			if snapshot, exists := v.snapshots[startEpoch]; exists {
+				cache[startEpoch] = snapshot
 			}
 
-			block--
+			startEpoch--
 		}
 
 		v.snapshots = cache
 
-		return v.state.cleanValidatorSnapshotsFromDB(latestBlock)
+		return v.state.cleanValidatorSnapshotsFromDB(latestEpoch)
 	}
 
 	return nil
 }
+
+// getLastCachedSnapshot gets the latest snapshot cached
+// If it doesn't have snapshot cached for desired epoch, it will return the latest one it has
+func (v *validatorsSnapshotCache) getLastCachedSnapshot(currentEpoch uint64) (*validatorSnapshot, error) {
+	snapshot := v.snapshots[currentEpoch]
+	if snapshot != nil {
+		v.logger.Trace("Found snapshot in memory cache", "Epoch", currentEpoch)
+
+		return snapshot, nil
+	}
+
+	snapshot, err := v.state.getLastSnapshot()
+	if err != nil {
+		return nil, err
+	}
+
+	return snapshot, nil
+}
+
+// getNextEpochEndingBlock gets the epoch ending block of a newer epoch
+// It start checking the blocks from the provided epoch ending block of the previous epoch
+func (v *validatorsSnapshotCache) getNextEpochEndingBlock(latestEpochEndingBlock uint64) (uint64, error) {
+	blockNumber := latestEpochEndingBlock + 1 // get next block
+
+	_, extra, err := getBlockData(blockNumber, v.blockchain)
+	if err != nil {
+		return 0, err
+	}
+
+	startEpoch := extra.Checkpoint.EpochNumber
+	epoch := startEpoch
+
+	for startEpoch == epoch {
+		blockNumber++
+
+		_, extra, err = getBlockData(blockNumber, v.blockchain)
+		if err != nil {
+			if errors.Is(err, blockchain.ErrNoBlock) {
+				return blockNumber - 1, nil
+			}
+
+			return 0, err
+		}
+
+		epoch = extra.Checkpoint.EpochNumber
+	}
+
+	return blockNumber - 1, nil
+}
+
+// isEpochEndingBlock checks if given block is an epoch ending block
+func (v *validatorsSnapshotCache) isEpochEndingBlock(blockNumber uint64, extra *Extra) (bool, error) {
+	if !extra.Validators.IsEmpty() {
+		return true, nil
+	}
+
+	_, nextBlockExtra, err := getBlockData(blockNumber+1, v.blockchain)
+	if err != nil {
+		if errors.Is(err, blockchain.ErrNoBlock) {
+			return true, nil
+		}
+
+		return false, nil
+	}
+
+	return extra.Checkpoint.EpochNumber != nextBlockExtra.Checkpoint.EpochNumber, nil
```

Comment:
Please double check me if all the use cases are covered, but I would expect something like this for epoch-ending blocks check:

```suggestion
	_, nextExtra, err := getBlockData(blockNumber+1, v.blockchain)
	if err != nil {
		if errors.Is(err, blockchain.ErrNoBlock) {
			return false, nil
		}

		return false, err
	}

	return nextExtra.Checkpoint.EpochNumber > extra.Checkpoint.EpochNumber, nil
```

Reply:
Just needed to return an error there. Fix: https://github.com/0xPolygon/polygon-edge/pull/997/commits/6239de4e5f05874320d4906782c6294f293fa677